crates/chia-consensus/src/merkle_tree.rs

```diff
@@ -0,0 +1,982 @@
+// This file contains the code used to create a full MerkleSet and is heavily reliant on the code in merkle_set.rs.
+
+use crate::merkle_set::{get_bit, hash, NodeType, BLANK};
+use clvmr::sha2::{Digest, Sha256};
+#[cfg(feature = "py-bindings")]
+use pyo3::exceptions;
+#[cfg(feature = "py-bindings")]
+use pyo3::prelude::*;
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyBytes, PyList};
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+
+// the ArrayTypes  used to create a more lasting MerkleTree representation in the MerkleSet struct
+#[derive(PartialEq, Debug, Copy, Clone)]
+pub enum ArrayTypes {
+    Leaf { data: u32 },              // indexes for the data_hash array
+    Middle { children: (u32, u32) }, // indexes for a Vec of ArrayTypes
+    Empty,
+    Truncated,
+}
+
+// represents a MerkleTree by putting all the nodes in a vec. Root is the last entry.
+#[derive(PartialEq, Debug, Clone, Default)]
+#[cfg_attr(feature = "py-bindings", pyclass(frozen, name = "MerkleSet"))]
+pub struct MerkleSet {
+    nodes_vec: Vec<ArrayTypes>,
+    leaf_vec: Vec<[u8; 32]>,
+    hash_cache: Vec<[u8; 32]>, // same size and order as nodes_vec
+}
+
+const EMPTY: u8 = 0;
+const TERMINAL: u8 = 1;
+const TRUNCATED: u8 = 2;
+const MIDDLE: u8 = 3;
+
+const EMPTY_NODE_HASH: [u8; 32] = [
+    127, 156, 158, 49, 172, 130, 86, 202, 47, 37, 133, 131, 223, 38, 45, 188, 125, 111, 104, 242,
+    160, 48, 67, 213, 201, 154, 74, 229, 167, 57, 108, 233,
+];
+// the above was calculated from the following code snippet which was previously used
+/*
+    let mut hasher = Sha256::new();
+    hasher.update([NodeType::Empty as u8]);
+    hasher.update(BLANK);
+    hasher.finalize().into();
+*/
+
+#[derive(Debug)]
+#[cfg_attr(feature = "py-bindings", pyclass(frozen, name = "SetError"))]
+pub struct SetError;
+
+pub fn deserialize_proof(proof: &[u8]) -> Result<MerkleSet, SetError> {
+    let mut merkle_tree = MerkleSet::default();
+    let pos = deserialize_recurse(&mut merkle_tree, proof, 0)?;
+    if pos != proof.len() {
+        Err(SetError)
+    } else {
+        Ok(merkle_tree)
+    }
+}
+
+fn deserialize_recurse(
+    merkle_tree: &mut MerkleSet,
+    proof: &[u8],
+    pos: usize,
```

Comment:
I think it would be better to pass in a sub-slice of `proof` when recursing, and not pass `pos` at all. The return value could then be the number of bytes consumed by the call, and you add it up as you return from the recursion